pymc3/variational/approximations.py

```diff
@@ -0,0 +1,474 @@
+
+        Returns
+        -------
+        global posterior space
+        """
+        raise NotImplementedError
+
+    def posterior_local(self, initial):
+        """Implements posterior distribution from initial latent space
+
+        Parameters
+        ----------
+        initial : initial latent space shape(samples, size)
+
+        Returns
+        -------
+        local posterior space
+        """
+        if not self.local_vars:
+            return initial
+        mu, rho = self._local_mu_rho()
+        x = initial * rho2sd(rho) + mu
+        return x
+
+    def _local_mu_rho(self):
+        mu = []
+        rho = []
+        for var in self.local_vars:
+            mu.append(self.known[var][0].ravel())
+            rho.append(self.known[var][1].ravel())
+        mu = tt.concatenate(mu)
+        rho = tt.concatenate(rho)
+        return mu, rho
+
+    def to_flat_input(self, node):
+        """Replaces vars with flattened view stored in self.input
+        """
+        return theano.clone(node, self.flat_view.replacements, strict=False)
+
+    def log_q_W_local(self, posterior):
+        """log_q_W samples over q for local vars
+
+        Gradient wrt mu, rho in density parametrization
+        is set to zero to lower variance of ELBO
+        """
+        if not self.local_vars:
+            return 0
+        logp = []
+        for var in self.local_vars:
+            mu = self.known[var][0].ravel()
+            rho = self.known[var][1].ravel()
+            q = log_normal3(self.view_from(posterior, var.name, 'local'), Z(mu), Z(rho))
+            logp.append(self.weighted_logp(var, q))
+        samples = tt.sum(tt.concatenate(logp, axis=1), axis=1)
+        return samples
+
+    def log_q_W_global(self, posterior):
+        """log_q_W samples over q for global vars
+        """
+        raise NotImplementedError
+
+    def log_q_W(self, posterior):
+        """log_q_W samples over q
+        """
+        return self.log_q_W_global(posterior) + self.log_q_W_local(posterior)
+
+    def log_p_D(self, posterior):
+        """log_p_D samples over q
+        """
+        _log_p_D_ = tt.sum(
+            list(map(self.weighted_logp, self.model.observed_RVs))
+        )
+        _log_p_D_ = self.to_flat_input(_log_p_D_)
+        samples = self.sample_over_space(posterior, _log_p_D_)
+        return samples
+
+    def weighted_logp(self, var, logp=None):
+        """Weight logp according to given population size
+        """
+        tot = self.population.get(var)
+        if logp is None:
+            logp = tt.sum(var.logpt)
+        if tot is not None:
+            tot = tt.as_tensor(tot)
+            logp *= tot / var.shape[0]
+        return logp
+
+    def KL_q_p_W(self, posterior):
+        """KL(q||p) samples over q
+        """
+        return self.log_q_W(posterior) - self.log_p_W(posterior)
+
+    def log_p_W_local(self, posterior):
+        if not self.local_vars:
+            return 0
+        _log_p_W_local_ = tt.sum(
```

Comment:
In the current version of code you can set the population for local variable and just use weights that sum up to 1. You always know the population as it is equal to population of decoded observed variable. There is no need in `M` any more.